File: src/lib/backtest/engine.ts
/**
 * Deterministic candle-replay backtester.
 *
 * Walks a strategy bar-by-bar over historical candles with a fixed
 * starting equity. One position open at a time (BTC-only v1).
 * Intra-bar stop/target checks assume pessimistic ordering: if both
 * levels are crossed within the same bar, the stop wins for longs
 * (a real bar could have traded stop-first) and the target wins for
 * shorts — symmetric unfavorable assumption.
 */
import { RingBuffer, makeIndicatorCache } from "../signals/ring-buffer";
import { aggregate, type EnsembleOptions } from "../signals/ensemble";
import type { Candle, Strategy, StrategyContext } from "../signals/types";
import { computeMetrics, type EquityPoint, type Metrics, type TradeRecord } from "./metrics";

export interface BacktestOptions {
  candles: readonly Candle[];
  strategies: readonly Strategy[];
  ensemble: EnsembleOptions;
  symbol?: string;
  startEquity?: number;
  riskPerTrade?: number;  // fraction of equity risked per trade (e.g. 0.02)
  slippageBps?: number;
  feeBps?: number;
  bufferSize?: number;
  barMinutes?: number;
}

export interface BacktestResult {
  metrics: Metrics;
  trades: TradeRecord[];
  equity: EquityPoint[];
  signals: Array<{ ts: number; strategyId: string; side: string; score: number; confidence: number }>;
}

interface OpenPosition {
  side: "long" | "short";
  entryTs: number;
  entryPrice: number;
  qty: number;
  stop: number;
  target: number;
}

export function runBacktest(opts: BacktestOptions): BacktestResult {
  const {
    candles,
    strategies,
    ensemble,
    symbol = "BTC-USD",
    startEquity = 25_000,
    riskPerTrade = 0.02,
    slippageBps = 1,
    feeBps = 5,
    bufferSize = 500,
    barMinutes = 1,
  } = opts;

  const buffer = new RingBuffer<Candle>(bufferSize);
  const indicators = makeIndicatorCache();
  let equity = startEquity;
  let position: OpenPosition | null = null;
  const trades: TradeRecord[] = [];
  const equitySeries: EquityPoint[] = [];
  const signalLog: BacktestResult["signals"] = [];

  const slip = slippageBps / 10_000;
  const fee = feeBps / 10_000;

  for (let i = 0; i < candles.length; i++) {
    const bar = candles[i];
    buffer.push(bar);
    indicators.clear();

    // Intra-bar exit checks BEFORE new signal (the bar has already traded)
    if (position) {
      const exitInfo = checkIntrabar(bar, position);
      if (exitInfo) {
        const fillPx = applySlippage(exitInfo.price, position.side === "long" ? "sell" : "buy", slip);
        const pnl = realizedPnl(position, fillPx, fee);
        equity += pnl;
        trades.push({
          entryTs: position.entryTs,
          exitTs: bar.ts,
          side: position.side,
          entryPrice: position.entryPrice,
          exitPrice: fillPx,
          qty: position.qty,
          pnl,
          reason: exitInfo.reason,
        });
        position = null;
      }
    }

    // Evaluate strategies on this bar
    const ctx: StrategyContext = {
      now: bar.ts,
      symbol,
      candles: buffer,
      ticks: new RingBuffer(1),
      indicators,
      params: {},
    };
    const signals = strategies.map((s) => {
      const sig = s.onCandle(ctx);
      if (sig) signalLog.push({ ts: sig.ts, strategyId: sig.strategyId, side: sig.side, score: sig.score, confidence: sig.confidence });
      return sig;
    });
    const decision = aggregate(signals, ensemble);

    // Signal-driven exit (flat or reversal)
    if (position) {
      const wantFlat =
        decision.side === "flat" ||
        (position.side === "long" && decision.side === "short") ||
        (position.side === "short" && decision.side === "long");
      if (wantFlat) {
        const fillPx = applySlippage(bar.c, position.side === "long" ? "sell" : "buy", slip);
        const pnl = realizedPnl(position, fillPx, fee);
        equity += pnl;
        trades.push({
          entryTs: position.entryTs,
          exitTs: bar.ts,
          side: position.side,
          entryPrice: position.entryPrice,
          exitPrice: fillPx,
          qty: position.qty,
          pnl,
          reason: "signal",
        });
        position = null;
      }
    }

    // Entry
    if (!position && (decision.side === "long" || decision.side === "short")) {
      const hint = signals.find((s) => s?.side === decision.side && s?.entryHint)?.entryHint;
      const entry = applySlippage(bar.c, decision.side === "long" ? "buy" : "sell", slip);
      const stop = hint?.stop ?? (decision.side === "long" ? entry * 0.98 : entry * 1.02);
      const target = hint?.target ?? (decision.side === "long" ? entry * 1.04 : entry * 0.96);
      const riskPerUnit = Math.abs(entry - stop);
      if (riskPerUnit > 0) {
        const qty = (equity * riskPerTrade) / riskPerUnit;
        position = {
          side: decision.side === "long" ? "long" : "short",
          entryTs: bar.ts,
          entryPrice: entry,
          qty,
          stop,
          target,
        };
        equity -= qty * entry * fee; // entry fee
      }
    }

    // Mark-to-market equity snapshot
    const mtm = position ? markEquity(equity, position, bar.c) : equity;
    equitySeries.push({ ts: bar.ts, equity: mtm });
  }

  // Final flatten at end of series
  if (position && candles.length > 0) {
    const last = candles[candles.length - 1];
    const fillPx = applySlippage(last.c, position.side === "long" ? "sell" : "buy", slip);
    const pnl = realizedPnl(position, fillPx, fee);
    equity += pnl;
    trades.push({
      entryTs: position.entryTs,
      exitTs: last.ts,
      side: position.side,
      entryPrice: position.entryPrice,
      exitPrice: fillPx,
      qty: position.qty,
      pnl,
      reason: "eod",
    });
    equitySeries[equitySeries.length - 1].equity = equity;
    position = null;
  }

  const metrics = computeMetrics(equitySeries, trades, startEquity, barMinutes);
  return { metrics, trades, equity: equitySeries, signals: signalLog };
}

function checkIntrabar(
  bar: Candle,
  pos: OpenPosition,
): { price: number; reason: "stop" | "target" } | null {
  if (pos.side === "long") {
    if (bar.l <= pos.stop) return { price: pos.stop, reason: "stop" };
    if (bar.h >= pos.target) return { price: pos.target, reason: "target" };
  } else {
    if (bar.h >= pos.stop) return { price: pos.stop, reason: "stop" };
    if (bar.l <= pos.target) return { price: pos.target, reason: "target" };
  }
  return null;
}

function applySlippage(price: number, side: "buy" | "sell", slip: number): number {
  return side === "buy" ? price * (1 + slip) : price * (1 - slip);
}

function realizedPnl(pos: OpenPosition, exitPx: number, fee: number): number {
  const gross = pos.side === "long" ? (exitPx - pos.entryPrice) * pos.qty : (pos.entryPrice - exitPx) * pos.qty;
  const exitFee = pos.qty * exitPx * fee;
  return gross - exitFee;
}

function markEquity(cash: number, pos: OpenPosition, mark: number): number {
  const unrealized = pos.side === "long" ? (mark - pos.entryPrice) * pos.qty : (pos.entryPrice - mark) * pos.qty;
  return cash + unrealized;
}

File: python/backtester/engine.py
"""Vectorized backtest engine mirroring src/lib/backtest/engine.ts.

Walks bar-by-bar (not fully vectorized) so intra-bar stop/target checks
and ensemble ordering match the TS reference. Python's advantage is in
parameter sweeps via joblib, not per-run speed.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Iterable

import numpy as np
import pandas as pd


MIN_PER_YEAR = 365 * 24 * 60


@dataclass
class Trade:
    entry_ts: int
    exit_ts: int
    side: str
    entry_price: float
    exit_price: float
    qty: float
    pnl: float
    reason: str


@dataclass
class BacktestResult:
    equity: pd.DataFrame
    trades: list[Trade] = field(default_factory=list)
    metrics: dict[str, float] = field(default_factory=dict)


def _apply_slippage(price: float, side: str, slip: float) -> float:
    return price * (1 + slip) if side == "buy" else price * (1 - slip)


def _realized_pnl(pos: dict, exit_px: float, fee: float) -> float:
    if pos["side"] == "long":
        gross = (exit_px - pos["entry_price"]) * pos["qty"]
    else:
        gross = (pos["entry_price"] - exit_px) * pos["qty"]
    return gross - pos["qty"] * exit_px * fee


def run_backtest(
    candles: pd.DataFrame,
    strategy_signals: dict[str, pd.DataFrame],
    weights: dict[str, float],
    *,
    start_equity: float = 25_000,
    risk_per_trade: float = 0.02,
    slippage_bps: float = 1,
    fee_bps: float = 5,
    bar_minutes: int = 1,
    long_threshold: float = 0.25,
    short_threshold: float = 0.25,
) -> BacktestResult:
    """
    candles: DataFrame with columns ts (ms), o, h, l, c, v.
    strategy_signals: dict of {strategy_id: DataFrame with columns
        side, score, confidence, stop, target}, indexed same as candles.
    """
    slip = slippage_bps / 10_000
    fee = fee_bps / 10_000
    equity = start_equity
    position: dict | None = None
    trades: list[Trade] = []
    eq_rows: list[tuple[int, float]] = []

    total_weight = sum(weights.values())
    ids = list(weights.keys())

    for i, row in candles.iterrows():
        ts, o, h, l, c = int(row.ts), row.o, row.h, row.l, row.c

        # Intra-bar stop/target
        if position:
            exit_info = None
            if position["side"] == "long":
                if l <= position["stop"]:
                    exit_info = (position["stop"], "stop")
                elif h >= position["target"]:
                    exit_info = (position["target"], "target")
            else:
                if h >= position["stop"]:
                    exit_info = (position["stop"], "stop")
                elif l <= position["target"]:
                    exit_info = (position["target"], "target")
            if exit_info:
                px, reason = exit_info
                fill = _apply_slippage(px, "sell" if position["side"] == "long" else "buy", slip)
                pnl = _realized_pnl(position, fill, fee)
                equity += pnl
                trades.append(Trade(position["entry_ts"], ts, position["side"], position["entry_price"], fill, position["qty"], pnl, reason))
                position = None

        # Aggregate ensemble decision
        weighted_score = 0.0
        weighted_conf = 0.0
        t_weight = 0.0
        any_long_hint = None
        any_short_hint = None
        for sid in ids:
            sdf = strategy_signals[sid]
            side = sdf.at[i, "side"] if i in sdf.index else "null"
            if side in ("null", None) or (isinstance(side, float) and np.isnan(side)):
                continue
            w = weights[sid]
            conf = sdf.at[i, "confidence"]
            score = sdf.at[i, "score"]
            dir_mult = 1 if side == "long" else -1 if side == "short" else 0
            weighted_score += w * score * conf * (1 if dir_mult != 0 else 0)
            weighted_conf += w * conf
            t_weight += w
            if side == "long" and any_long_hint is None:
                any_long_hint = (sdf.at[i, "stop"], sdf.at[i, "target"])
            if side == "short" and any_short_hint is None:
                any_short_hint = (sdf.at[i, "stop"], sdf.at[i, "target"])
        if t_weight == 0:
            decision = "flat"
            norm_score = 0.0
        else:
            norm_score = weighted_score / t_weight
            decision = "long" if norm_score >= long_threshold else "short" if norm_score <= -short_threshold else "flat"

        # Signal-driven exit
        if position and (
            decision == "flat"
            or (position["side"] == "long" and decision == "short")
            or (position["side"] == "short" and decision == "long")
        ):
            fill = _apply_slippage(c, "sell" if position["side"] == "long" else "buy", slip)
            pnl = _realized_pnl(position, fill, fee)
            equity += pnl
            trades.append(Trade(position["entry_ts"], ts, position["side"], position["entry_price"], fill, position["qty"], pnl, "signal"))
            position = None

        # Entry
        if position is None and decision in ("long", "short"):
            hint = any_long_hint if decision == "long" else any_short_hint
            entry = _apply_slippage(c, "buy" if decision == "long" else "sell", slip)
            if hint and not np.isnan(hint[0]) and not np.isnan(hint[1]):
                stop, target = hint
            else:
                stop = entry * 0.98 if decision == "long" else entry * 1.02
                target = entry * 1.04 if decision == "long" else entry * 0.96
            risk_unit = abs(entry - stop)
            if risk_unit > 0:
                qty = (equity * risk_per_trade) / risk_unit
                position = {"side": decision, "entry_ts": ts, "entry_price": entry, "qty": qty, "stop": stop, "target": target}
                equity -= qty * entry * fee

        # MtM equity
        if position:
            mtm = equity + (
                (c - position["entry_price"]) * position["qty"]
                if position["side"] == "long"
                else (position["entry_price"] - c) * position["qty"]
            )
        else:
            mtm = equity
        eq_rows.append((ts, mtm))

    # Final flatten
    if position and len(candles):
        last = candles.iloc[-1]
        fill = _apply_slippage(last.c, "sell" if position["side"] == "long" else "buy", slip)
        pnl = _realized_pnl(position, fill, fee)
        equity += pnl
        trades.append(Trade(position["entry_ts"], int(last.ts), position["side"], position["entry_price"], fill, position["qty"], pnl, "eod"))
        eq_rows[-1] = (eq_rows[-1][0], equity)

    equity_df = pd.DataFrame(eq_rows, columns=["ts", "equity"])
    return BacktestResult(equity=equity_df, trades=trades, metrics=_metrics(equity_df, trades, start_equity, bar_minutes))


def _metrics(equity: pd.DataFrame, trades: list[Trade], start_equity: float, bar_minutes: int) -> dict[str, float]:
    if len(equity) < 2:
        return {"total_return_pct": 0, "sharpe": 0, "sortino": 0, "max_drawdown_pct": 0, "hit_rate": 0, "trades": 0, "final_equity": start_equity}
    final = float(equity.equity.iloc[-1])
    returns = equity.equity.pct_change().dropna().to_numpy()
    mean_r = returns.mean() if len(returns) else 0
    sd = returns.std(ddof=0) if len(returns) else 0
    downside = np.sqrt((np.where(returns < 0, returns, 0) ** 2).mean()) if len(returns) else 0
    annualizer = np.sqrt(MIN_PER_YEAR / bar_minutes)
    sharpe = 0 if sd == 0 else mean_r / sd * annualizer
    sortino = 0 if downside == 0 else mean_r / downside * annualizer
    peak = equity.equity.cummax()
    dd = (peak - equity.equity) / peak
    max_dd = float(dd.max())
    wins = sum(1 for t in trades if t.pnl > 0)
    hit = wins / len(trades) if trades else 0
    return {
        "total_return_pct": (final - start_equity) / start_equity * 100,
        "sharpe": float(sharpe),
        "sortino": float(sortino),
        "max_drawdown_pct": max_dd * 100,
        "hit_rate": hit,
        "trades": len(trades),
        "final_equity": final,
        "start_equity": start_equity,
    }


def eval_strategies(candles: pd.DataFrame, strategies: Iterable[Any]) -> dict[str, pd.DataFrame]:
    out: dict[str, pd.DataFrame] = {}
    h = candles.h.to_numpy()
    l = candles.l.to_numpy()
    c = candles.c.to_numpy()
    for s in strategies:
        res = s.evaluate(h, l, c)
        out[s.id] = pd.DataFrame(res, index=candles.index)
    return out
